CommonMark/ExtensibleHtmlPrinter.cs

```diff
@@ -0,0 +1,742 @@
+﻿using CommonMark.Syntax;
+using System.Collections.Generic;
+using System.Globalization;
+using System.IO;
+using System.Text;
+
+namespace CommonMark.Formatter
+{
+    public class ExtensibleHtmlPrinter : IPrinter
+    {
+        private static readonly char[] EscapeHtmlCharacters = new[] { '&', '<', '>', '"' };
+        private const string HexCharacters = "0123456789ABCDEF";
+
+        private static readonly char[] EscapeHtmlLessThan = "&lt;".ToCharArray();
+        private static readonly char[] EscapeHtmlGreaterThan = "&gt;".ToCharArray();
+        private static readonly char[] EscapeHtmlAmpersand = "&amp;".ToCharArray();
+        private static readonly char[] EscapeHtmlQuote = "&quot;".ToCharArray();
+
+        private static readonly string[] HeaderOpenerTags = new[] { "<h1>", "<h2>", "<h3>", "<h4>", "<h5>", "<h6>" };
+        private static readonly string[] HeaderCloserTags = new[] { "</h1>", "</h2>", "</h3>", "</h4>", "</h5>", "</h6>" };
+
+        private static readonly bool[] UrlSafeCharacters = new[] {
+            false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false,
+            false, false, false, false, false, false, false, false, false, false, false, false, false, false, false, false,
+            false, true,  false, true,  true,  true,  false, false, true,  true,  true,  true,  true,  true,  true,  true,
+            true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  false, true,  false, true,
+            true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,
+            true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  false, false, false, false, true,
+            false, true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,
+            true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  true,  false, false, false, false, false,
+        };
+
```

Comment:
All these fields should not be duplicated. I think making the `EscapeUrl`/`EscapeHtml` methods from `HtmlPrinter` internal should suffice.